### ISSUE TITLE
[Bug] Apple Metal/MPS -- TVM/MLC-LLM won't compile from source

### ISSUE THREAD

**Issue by BuildBackBuehler** [mlc-ai/mlc-llm#2540]
## To Reproduce

I've compiled each a few times each. But since I updated and attempted to compile, I've been unable to (*except once, not sure if pure luck or a matter of a stock, no features, build). With that little detail of success, it was off a fresh git repo DL, whereas when I have dropped the features after a failed build back to stock, it still fails. 

Features that seem to exacerbate the issue: BLAS, MKL, CoreML, Arm Compute Lib., basically anything that'd go through MPS and it causes this foundational error (as in there may be an error about inability to find, for the file src/runtime/contrib/ACL/allocator.cc, <#include acl/runtime/IAllocator.h> + Core/Types.h (which doesn't make sense, I've gone out of the way to incorporate the precise directory that IAlloc. is in my Include flags/CMake conf. (ACL/arm_compute/core + ACL/arm_compute/runtime):

```
[ 53%] Building CXX object tvm/CMakeFiles/tvm_objs.dir/src/tir/analysis/verify_ssa.cc.o
/Users/zack/.home/gitrepos/LLMLife/backend/tvm-unity/src/runtime/contrib/mps/conv.mm:36:25: error: 'CopyDataFromTo' is a protected member of 'tvm::runtime::metal::MetalWorkspace'
   36 |   entry_ptr->metal_api->CopyDataFromTo((__bridge void*)mtlbuf, 0, (__bridge void*)temp, 0,
      |                         ^
/Users/zack/.home/gitrepos/LLMLife/backend/tvm-unity/src/runtime/contrib/mps/../../metal/metal_common.h:187:8: note: declared protected here
  187 |   void CopyDataFromTo(const void* from, size_t from_size, void* to, size_t to_size, size_t size,
      |        ^
/Users/zack/.home/gitrepos/LLMLife/backend/tvm-unity/src/runtime/contrib/mps/conv.mm:72:25: error: 'CopyDataFromTo' is a protected member of 'tvm::runtime::metal::MetalWorkspace'
   72 |   entry_ptr->metal_api->CopyDataFromTo((__bridge void*)temp, 0, (__bridge void*)mtlbuf, 0,
      |                         ^
/Users/zack/.home/gitrepos/LLMLife/backend/tvm-unity/src/runtime/contrib/mps/../../metal/metal_common.h:187:8: note: declared protected here
  187 |   void CopyDataFromTo(const void* from, size_t from_size, void* to, size_t to_size, size_t size,
      |        ^
/Users/zack/.home/gitrepos/LLMLife/backend/tvm-unity/src/runtime/contrib/mps/conv.mm:106:53: error: no member named 'GetCommandQueue' in 'tvm::runtime::metal::MetalWorkspace'
  106 |   id<MTLCommandQueue> queue = entry_ptr->metal_api->GetCommandQueue(data->device);
      |                               ~~~~~~~~~~~~~~~~~~~~  ^
/Users/zack/.home/gitrepos/LLMLife/backend/tvm-unity/src/runtime/contrib/mps/conv.mm:115:25: error: 'CopyDataFromTo' is a protected member of 'tvm::runtime::metal::MetalWorkspace'
  115 |   entry_ptr->metal_api->CopyDataFromTo((__bridge void*)bufB, 0, (__bridge void*)tempB, 0,
      |                         ^
/Users/zack/.home/gitrepos/LLMLife/backend/tvm-unity/src/runtime/contrib/mps/../../metal/metal_common.h:187:8: note: declared protected here
  187 |   void CopyDataFromTo(const void* from, size_t from_size, void* to, size_t to_size, size_t size,
      |        ^
4 errors generated.
```

I originally posted about this issue in TVM 6 days ago...but it is inactive. https://github.com/mlc-ai/relax/issues/321
More errors/gen. information for context if needed. 

## Environment

 - Platform (e.g. WebGPU/Vulkan/IOS/Android/CUDA): Metal/MPS
 - Operating system (e.g. Ubuntu/Windows/MacOS/...): Mac V. 14.5
 - Device (e.g. iPhone 12 Pro, PC+RTX 3090, ...) MBP M1 Max
 - How you installed MLC-LLM (`conda`, source): Source (Github)
 - How you installed TVM-Unity (`pip`, source): Source (Github)
 - Python version (e.g. 3.10): 3.11.9
 - GPU driver version (if applicable):
 - CUDA/cuDNN version (if applicable):
 - TVM Unity Hash Tag (`python -c "import tvm; print('\n'.join(f'{k}: {v}' for k, v in tvm.support.libinfo().items()))"`, applicable if you compile models): [Haven't actually reinstalled TVM as a Python instance, just built. 
 - Any other relevant information: 

## Additional context

Sadly couldn't find anything on the net about how to fix this error. Figured there'd be a lot of these MTLCommandQueue errors but nothing concrete. 

I only tried compiling MLC-LLM 1 or 2 times. And that was with my TVM (that compiled that one time...well the .dylibs, but not the pure 100% instance, I had attempted more builds after). I suppose I'll give a try to compile MLC w/ 3rd Party TVM, but I need to manipulate MLC's quantization file so I can import a custom-quantized model of mine.

**Comment by tqchen** [comment]
Likely you don't need to turn on arm compute and mps since we generate our own metal code

**Comment by tqchen** [comment]
ah, you need to write device="metal"

**Comment by BuildBackBuehler** [comment]
> ah, you need to write device="metal"

🤦‍♂️ it's always something so stupid, agh! Thank you

Welp, ```ValueError: Cannot detect device `metal(0)`. Please make sure the device and its driver is installed properly, and TVM is compiled with the driver``` I'm guessing when I did that compilation that worked, I must've neglected to include Metal even. Will see what happens when I do another clean-build + Metal.

And while I'm here, I neglected to mention the warning that always serve as a precursor to the 4 "foundational" errors I had mentioned. 

```
[ 16%] Building CXX object CMakeFiles/tvm_runtime_objs.dir/src/runtime/workspace_pool.cc.o
/Users/zack/.home/gitrepos/LLMLife/backend/tvm-unity/src/runtime/threading_backend.cc:296:30: warning: zero as null pointer constant [-Wzero-as-null-pointer-constant]
  296 |     SetThreadFullCpuAffinity(CURRENT_THREAD_HANDLE, mode);
      |                              ^
/Users/zack/.home/gitrepos/LLMLife/backend/tvm-unity/src/runtime/threading_backend.cc:51:77: note: expanded from macro 'CURRENT_THREAD_HANDLE'
   51 | #define CURRENT_THREAD_HANDLE (static_cast<std::thread::native_handle_type>(0))
      |                                                                             ^
/Users/zack/.home/gitrepos/LLMLife/backend/tvm-unity/src/runtime/threading_backend.cc:441:25: warning: zero as null pointer constant [-Wzero-as-null-pointer-constant]
  441 |       SetThreadAffinity(CURRENT_THREAD_HANDLE,
      |                         ^
/Users/zack/.home/gitrepos/LLMLife/backend/tvm-unity/src/runtime/threading_backend.cc:51:77: note: expanded from macro 'CURRENT_THREAD_HANDLE'
   51 | #define CURRENT_THREAD_HANDLE (static_cast<std::thread::native_handle_type>(0))
```

**Comment by BuildBackBuehler** [comment]
[I have a feeling this is the answer to my prayers. ](https://github.com/Genivia/ugrep/issues/305https://github.com/Genivia/ugrep/issues/305) 

I had swapped out DMLC_Core and the related files that were changed when that was updated without luck. Then sought to check out the last time those threading/pool files were modified. 

Haven't tried those files yet but figure it must be Apple's policies w/ threading and [this seems to confirm](https://github.com/medooze/media-server/issues/63)

But donno, just glazed over and saw a couple apples but maybe compared to a couple oranges.

**Comment by BuildBackBuehler** [comment]
I think this change to runtime/threading_backend.cc was what did it based on the warnings pre-error. Nothing in MPS has changed in years, Metal, a bit changed 3 weeks back, haven't looked into whether or not that helps/hurts/insignificant.

// This global function can be used by disco runtime to bind processes
// to CPUs.
TVM_REGISTER_GLOBAL("tvm.runtime.threading.set_current_thread_affinity")
    .set_body_typed([](IntTuple cpu_ids) {
      SetThreadAffinity(CURRENT_THREAD_HANDLE,
                        std::vector<unsigned int>{cpu_ids.begin(), cpu_ids.end()});
    });

**Comment by BuildBackBuehler** [comment]
I was able to compile TVM/MLC but its producing segmentation fault errors on conversion (weights) of my Codestral model. Also gotten errors with compiling a 3-bit Omniquant Llama model (gen_config worked fine) and trying to chat with an AQLM 2-bit model I managed to get to compile previously. However, I'm not sure if the .dylib I'd compiled was legitimate (used no_quant for gen_config/compilation) so I'd need to double back on another compile anyways. 

Seems it is the fact that there's only a 

"Protected: CopyDataFromTo(vars, etc. etc.)"
  
No "Public: CopyDataFromTo" defined in runtime/metal/metal_common.h. There's also no "GetCommandQueue" (used in metal_api.mm and conv.mm (IIRC)) defined in metal_common.h. 

I remedied those issues. Then it was only a matter of introducing PublicCopyDataFromTo function & the GCQ definition in conv.mm, gemm.mm (contrib/MPS files) and metal_api.mm. Well, if the resolution is kosher, but I guess from the errors I'm experiencing, it breaks something important (I'm guessing the data *should* be transported in a protected state. I'm thinking that there must've just been a discrepancy (in conv.mm or metal_api?)/missing definition (GCQ) prohibiting it from protected data transference

**Comment by BuildBackBuehler** [comment]
[1]    63036 segmentation fault  mlc_llm convert_weight /Users/zack/.home/local/models/Uncensored_Llama-70B
/Users/zack/.home/local/mise/installs/python/3.11.9/lib/python3.11/multiprocessing/resource_tracker.py:254: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '

I'm losing my sanity here at this point. My Python/Poetry appear to be ARM64...so it nixes that possibility. I checked because I saw all the multiprocessing errors here have been related to that or other user errors. 

My last remaining guess, and I wish I just turned it off when I turned off ARM Comp. Lib. is BLAS. I'm feeling a bit dumb now really, because I believe when issues started and I posted this in TVM, I noted Apple BLAS was suspect. I think the code is out of date because it shoots warnings (and before I messed around w/ the Metal/MPS code, errors!) about how the code is relying on sgemm and dgemm or w/e functions/scripts that are deprecated. And I tried everything under the sun to force CMake to incorporate the new Apple BLAS without change. So I'll be turning that off now, too. 

I do have tons of modules on normally. AOTExec, UMA, Threads, RPC, CPP TVM, CPP RPC, Profiler, Graph Executor, CoreML, TCMalloc, MLIR, Pipeline. I think that _might_ be it 😂😅
